File: Backend/Core/Entities/BaseSalaryComposition.cs
using Core.Entities.Enums;

namespace Core.Entities
{
    /// <summary>
    /// Lớp cơ sở chứa các thuộc tính dùng chung cho cả TPL Hệ thống và TPL Người dùng
    /// </summary>
    public abstract class BaseSalaryComposition
    {
        public string ScCode { get; set; } = string.Empty;
        public string ScName { get; set; } = string.Empty;
        
        public SalaryCompositionType ScType { get; set; } 
        
        public SalaryCompositionNature ScNature { get; set; }
        public TaxStatus ScTaxStatus { get; set; }
        public bool ScIsTaxDeductible { get; set; } // tinyint (0/1) -> bool
        
        public string? ScTaxableExpression { get; set; } = string.Empty;
        public string? ScExemptExpression { get; set; } = string.Empty;
        public string? ScLimitExpression { get; set; } = string.Empty;
        public bool ScAllowExceedLimit { get; set; }
        
        public Enums.ValueType ScValueType { get; set; }
        public CalculationMethod ScCalculationMethod { get; set; }
        public AggregationScope ScAggregationScope { get; set; }
        
        public string ScCompositionCode { get; set; } = string.Empty;
        public int? ScOrganizationLevel { get; set; }
        
        public string? ScFormulaExpression { get; set; } = string.Empty;
        public string ScDescription { get; set; } = string.Empty;
        
        public DisplayOnPayroll ScIsDisplayedOnPayroll { get; set; }
        public bool ScIsDeleted { get; set; } // tinyint (0/1) -> bool
    }
}

File: Frontend/src/models/BaseSalaryComposition.js
/**
 * @file Lớp cơ sở cho Thành phần lương.
 * @author Phuong 2026-06-08
 */

class BaseSalaryComposition {
  constructor(data = {}) {
    this.ScCode = data.ScCode || ''
    this.ScName = data.ScName || ''
    this.ScType = data.ScType ?? 0
    this.ScNature = data.ScNature ?? 0
    this.TaxStatus = data.TaxStatus ?? 0
    this.IsTaxDeductible = data.IsTaxDeductible ?? false

    this.TaxableExpression = data.TaxableExpression || null
    this.ExemptExpression = data.ExemptExpression || null
    this.LimitExpression = data.LimitExpression || null
    this.AllowExceedLimit = data.AllowExceedLimit ?? false

    this.ValueType = data.ValueType ?? 0
    this.CalculationMethod = data.CalculationMethod ?? 0
    this.AggregationScope = data.AggregationScope ?? 0

    this.CompositionCode = data.CompositionCode || ''
    this.OrganizationLevel = data.OrganizationLevel || null

    this.FormulaExpression = data.FormulaExpression || null
    this.Description = data.Description || ''
    this.IsDisplayedOnPayroll = data.IsDisplayedOnPayroll ?? 0
  }
}

export default BaseSalaryComposition
